pkg/twcc/twcc.go

```diff
@@ -33,68 +37,118 @@ type Recorder struct {
+
+		// NOTE: we don't erase packets from the history in case they need to be resent
+		// after a reordering. They will be removed instead in Record when they get too
+		// old.
+	}
+	return feedbacks
+}
+
+// maybeBuildFeedbackPacket builds a feedback packet starting from startSN (inclusive) until
+// endSN (exclusive).
+func (r *Recorder) maybeBuildFeedbackPacket(beginSeqNumInclusive, endSeqNumExclusive int64) *feedback {
+	// NOTE: The logic of this method is inspired by the implementation in Chrome.
+	// See https://source.chromium.org/chromium/chromium/src/+/refs/heads/main:third_party/webrtc/modules/remote_bitrate_estimator/remote_estimator_proxy.cc;l=276;drc=b5cd13bb6d5d157a5fbe3628b2dd1c1e106203c6
+	startSN, endSN := r.arrivalTimeMap.Clamp(beginSeqNumInclusive), r.arrivalTimeMap.Clamp(endSeqNumExclusive)
+
+	// Create feedback on demand, as we don't yet know if there are packets in the range that have been
+	// received.
+	var fb *feedback
+
+	nextSequenceNumber := beginSeqNumInclusive
 
-	var pkts []rtcp.Packet
-	for _, pkt := range r.receivedPackets {
-		ok := feedback.addReceived(uint16(pkt.sequenceNumber&0xffff), pkt.arrivalTime)
-		if !ok {
-			pkts = append(pkts, feedback.getRTCP())
-			feedback = newFeedback(r.senderSSRC, r.mediaSSRC, r.fbPktCnt)
+	for seq := startSN; seq < endSN; seq++ {
+		foundSeq, arrivalTime, ok := r.arrivalTimeMap.FindNextAtOrAfter(seq)
+		seq = foundSeq
+		if !ok || seq >= endSN {
+			break
+		}
+
+		if fb == nil {
+			fb = newFeedback(r.senderSSRC, r.mediaSSRC, r.fbPktCnt)
 			r.fbPktCnt++
-			feedback.addReceived(uint16(pkt.sequenceNumber&0xffff), pkt.arrivalTime)
+
+			// It should be possible to add seq to this new packet.
+			// If the difference between seq and beginSeqNumInclusive is too large, discard
+			// reporting too old missing packets.
+			baseSequenceNumber := max64(beginSeqNumInclusive, seq-maxMissingSequenceNumbers)
+
+			// baseSequenceNumber is the expected first sequence number. This is known,
+			// but we may not have actually received it, so the base time should be the time
+			// of the first received packet in the feedback.
+			fb.setBase(uint16(baseSequenceNumber), arrivalTime)
+
+			if !fb.addReceived(uint16(seq), arrivalTime) {
+				// Could not add a single received packet to the feedback.
+				// This is unexpected to actually occur, but if it does, we'll
+				// try again after skipping any missing packets.
+				// NOTE: It's fine that we already incremented fbPktCnt, as in essence
+				// we did actually "skip" a feedback (and this matches Chrome's behavior).
+				r.startSequenceNumber = &seq
+				return nil
+			}
+		} else if !fb.addReceived(uint16(seq), arrivalTime) {
+			// Could not add timestamp. Packet may be full. Return
+			// and try again with a fresh packet.
+			break
 		}
+
+		nextSequenceNumber = seq + 1
 	}
-	r.receivedPackets = []pktInfo{}
-	pkts = append(pkts, feedback.getRTCP())
 
-	return pkts
+	r.startSequenceNumber = &nextSequenceNumber
+	return fb
```

Comment:
guess it is possible that some packets could be duplicated in feedback. But, I believe it should be fine. Just wanted to make sure that it is fine. For example, will the following send duplicate feedback?
- Packets 11, 12, 13 - feedback sent
- Packets 14, 15, 16, 10 - arrive -> 10 arriving would have set the start sequence number back to 10.
- Next feedback - 10, 11, 12, 13, 14, 15, 16 -> 11, 12 , 13 duplicated, but they should have the same data

Reply:
I believe libwebrtc expects duplicated feedback on reorderings: https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/remote_bitrate_estimator/remote_estimator_proxy_unittest.cc;l=340;bpv=0;bpt=0. I can't find anything in the [RFC](https://datatracker.ietf.org/doc/html/draft-holmer-rmcat-transport-wide-cc-extensions-01) that says one way or the other, so defaulting to libwebrtc's behavior seems reasonable here.